lib/galaxy/webapps/galaxy/controllers/root.py

```diff
@@ -83,6 +84,17 @@ def login(self, trans, redirect=None, **kwd):
         """
         User login path for client-side.
         """
+        # directly redirect to oidc provider if 1) enable_oidc is True, 2)
+        # there is only one oidc provider, 3) auth_conf.xml has no authenticators
+        if (trans.app.config.enable_oidc and
+                len(trans.app.config.oidc) == 1 and
+                len(trans.app.auth_manager.authenticators) == 0):
+
+            provider = trans.app.config.oidc[0]
+            success, message, redirect_uri = trans.app.authnz_manager.authenticate(provider, trans)
+            if success:
+                return trans.response.send_redirect(redirect_uri)
```

Comment:
this section has overlap with the controller, so I guess it should leverage it to avoid duplication (and possibly diverging logic).

Reply:
Hi @VJalili, thanks for reviewing. When you say "this section has overlap with the controller", which controller do you mean? Are you saying this logic should be in the authnz_manager?

Reply:
Sorry I was not clear, [this is the endpoint in the AuthNZ controller](https://github.com/galaxyproject/galaxy/blob/850d483bf260cba969ddb21de2f5404c15e09d1c/lib/galaxy/webapps/galaxy/controllers/authnz.py#L47-L49) that I was referring to. I wonder how we can leverage that controller hence to avoid re-implementing a similar logic in different places. (Having same logic implemented in different place can be problematic when we modify one and possibly forget the other one, therefore, I would recommend we leverage the aforelinked controller for calling the authnz manager.)

Reply:
It looks to me like this login method should ideally be retired soon anyway, and replaced with the login method in the authnz controller eventually? This looks like something that'll have to happen as part of retiring mako and shifting to Vue? However, it looks like the authnz manager doesn't support non-oidc login. https://github.com/galaxyproject/galaxy/blob/850d483bf260cba969ddb21de2f5404c15e09d1c/lib/galaxy/webapps/galaxy/controllers/authnz.py#L43
I was going to suggest removing that line in the authnz manager and if `enable_oidc` is false, it could redirect to the standard login page instead. Then, this method in root.py could call the authnz manager's login method. Would that work?

Reply:
Not sure I follow.  The route @VJalili links above requires a specific provider and handles the handshake, yes, but it doesn't present any of the interface.

We still want a first class route to /login that serves the (already Vue) login entrypoint.

Reply:
Sorry, I hadn't realized that the endpoint Vahid pointed to was actually provider specific, so no, what I said doesn't apply.

Reply:
@VJalili it's a worthy consideration and if you'd like to refactor this down the road, we can certainly do that.  Did you mean something like this?
```
        if (trans.app.config.enable_oidc and
                len(trans.app.config.oidc) == 1 and
                len(trans.app.auth_manager.authenticators) == 0):

            return trans.webapp.controllers['authnz'].login(trans, trans.app.config.oidc[0])
```
It doesn't feel to me like that really cleans up a lot, and without trying it it *looks* like you would end up in a weird state potentially without a page to look at, since it returns json?

Anyway, I do think this works for now and I'd like to go ahead and get it in, certainly feel free to follow up down the road though.

Reply:
I guess the snippet @dannon has mentioned could be helpful leveraging the controller instead of re-implementing its logic here, which we may easily forget updating it when making any changes to the controller. That brings me to a second point, would it be useful to implement a unit test for this functionality?